<!doctype html>
<html lang="en" class="h-100">
  <head>
  <meta name="generator" content="Hugo 0.53" />
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <link rel="stylesheet" href="https://nekrut.github.io/BMMB554/css/bootstrap.min.css">
  
  
  <title>Syllabus | BMMB554</title>
  <style>
.container {
  max-width: 800px;
}
#nav a {
  font-weight: bold;
  color: inherit;
}
#nav a.nav-link-active {
  background-color: #212529;
  color: #fff;
}
#nav-border {
  border-bottom: 1px solid #212529;
}
#main {
  margin-top: 1em;
  margin-bottom: 4em;
}
#home-jumbotron {
  background-color: inherit;
}
#footer .container {
  padding: 1em 0;
}
#footer a {
  color: inherit;
  text-decoration: underline;
}
.font-125 {
  font-size: 125%;
}
.tag-btn {
  margin-bottom: 0.3em;
}
pre {
  background-color: #f5f5f5;
  border: 1px solid #ccc;
  border-radius: 4px;
  padding: 16px;
}
pre code {
  padding: 0;
  font-size: inherit;
  color: inherit; 
  background-color: transparent;
  border-radius: 0;
}
code {
  padding: 2px 4px;
  font-size: 90%;
  color: #c7254e;
  background-color: #f9f2f4;
  border-radius: 4px;
}
img {
  max-width: 100%;
}
</style>

  <script type="text/javascript" async
  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
  MathJax.Hub.Config({
  tex2jax: {
    inlineMath: [['$','$'], ['\\(','\\)']],
    displayMath: [['$$','$$']],
    processEscapes: true,
    processEnvironments: true,
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
    TeX: { equationNumbers: { autoNumber: "AMS" },
         extensions: ["AMSmath.js", "AMSsymbols.js"] }
  }
  });
  MathJax.Hub.Queue(function() {
    
    
    
    var all = MathJax.Hub.getAllJax(), i;
    for(i = 0; i < all.length; i += 1) {
        all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });

  MathJax.Hub.Config({
  
  TeX: { equationNumbers: { autoNumber: "AMS" } }
  });
</script>
  
</head>
  <body class="d-flex flex-column h-100">
    <div id="nav-border" class="container">
  <nav id="nav" class="nav justify-content-center">
  
  
  
    
    
      
        
      
    
    
    <a class="nav-link " href="/BMMB554/"><i data-feather="home"></i> Home</a>
  
    
    
      
        
      
    
    
    <a class="nav-link " href="/BMMB554/post/"><i data-feather="book-open"></i> Lectures</a>
  
    
    
      
        
      
    
    
    <a class="nav-link " href="/BMMB554/tags/"><i data-feather="tag"></i> Tags</a>
  
    
    
      
       
        
        
      
    
    
    <a class="nav-link nav-link-active" href="/BMMB554/about/"><i data-feather="list"></i> Syllabus</a>
  
  </nav>
</div>
    <div class="container">
      <main id="main">
        

<h1>Syllabus</h1>


<p><a href="https://xkcd.com/451/"><img src="http://imgs.xkcd.com/comics/impostor.png" alt="" /></a></p>

<h2 id="place-and-time">Place and Time</h2>

<hr />

<p>Boucke Building 216 | Monday, Wednesday 2:30pm - 3:45pm</p>

<iframe src="https://www.google.com/maps/embed?pb=!1m14!1m8!1m3!1d12081.221080223779!2d-77.8618033!3d40.7992851!3m2!1i1024!2i768!4f13.1!3m3!1m2!1s0x0%3A0x83b4bcab5cf7d65e!2sBoucke+Building!5e0!3m2!1sen!2sus!4v1546882122664" width="400" height="200" frameborder="0" style="border:0" allowfullscreen></iframe>

<p><br></p>

<h2 id="instructor">Instructor</h2>

<hr />

<p><strong>Anton Nekrutenko</strong><br>
<a href="mailto:aun1@psu.edu?Subject=BMMB554">aun1@psu.edu</a><br>
Wartik 505<br>
Office hours by appointment only
<div class="alert alert-warning" role="alert">
  When contacting instructor use the above e-mail and include &ldquo;BMMB554&rdquo; in the subject line (simply click on e-mail address).
</div></p>

<h2 id="course-logistics">Course logistics</h2>

<hr />

<p>This course <strong>does not use</strong> Canvas. Canvas is a convoluted system with too many features and undefined purpose. Instead, this course is served from <a href="https://github.com/nekrut/BMMB554">GitHub</a>. Each new lecture will be created as a new entry under &ldquo;Lectures&rdquo; section of the website.</p>

<div class="alert alert-danger" role="alert">
  <b>Do not contact</b> me through Canvas! I will not check my Inbox there. Instead, contact me via e-mail as described above.
</div>

<h2 id="grading">Grading</h2>

<hr />

<p><div class="progress" style="height: 50px;">
  <div class="progress-bar bg-danger" role="progressbar" style="width: 25%" aria-valuenow="25" aria-valuemin="0" aria-valuemax="100">Attendance (25%)</div>
  <div class="progress-bar bg-success" role="progressbar" style="width: 30%" aria-valuenow="30" aria-valuemin="0" aria-valuemax="100">Homework (30%)</div>
  <div class="progress-bar bg-info" role="progressbar" style="width: 45%" aria-valuenow="45" aria-valuemin="0" aria-valuemax="100">Final Project (45%)</div>
</div>
<br>
A short homework will be given once a week. We will discuss project requirements before the spring break.</p>

<h2 id="topics">Topics</h2>

<hr />

<p>We will cover a breadth of topics. The course will be divided into several blocks:</p>

<h4 id="block-1-a-toolset-for-genomics"><strong>Block 1</strong>: A toolset for genomics</h4>






<table class="table table-hover table-primary">
<thead>
<tr>
<th>#</th>
<th>What</th>
<th>Why</th>
</tr>
</thead>

<tbody>
<tr>
<td>1.1</td>
<td><a href="/BMMB554/lecture1/"><strong>Introduction to this course</strong></a></td>
<td>Who you are and what you want to accomplish</td>
</tr>

<tr>
<td>1.2</td>
<td><a href="/BMMB554/lecture2/"><strong>Jupyter and GitHub</strong></a></td>
<td>Throughout this course Jupyter will be our primary framework for data analysis and GitHub will be used for version control</td>
</tr>

<tr>
<td>1.3</td>
<td><strong>Python <a href="/BMMB554/lecture3/">1</a> and <a href="/BMMB554/lecture4/">2</a></strong></td>
<td>Understanding transactional Python for data analysis</td>
</tr>

<tr>
<td>1.4</td>
<td><strong><a href="/BMMB554/lecture5/">NumPy and SciPy</a></strong></td>
<td>Libraries for numerical magic</td>
</tr>

<tr>
<td>1.5</td>
<td><strong>Pandas and SQL <a href="/BMMB554/lecture6/">1</a> and <a href="/BMMB554/lecture7">2</a></strong></td>
<td>Operating on large datasets in Python</td>
</tr>

<tr>
<td>1.6</td>
<td><a href="/BMMB554/lecture8"><strong>MatPlotLib &amp; Seaborn</strong></a></td>
<td>Plotting data in Python</td>
</tr>

<tr>
<td>1.7</td>
<td><a href="/BMMB554/lecture9"><strong>Galaxy</strong></a></td>
<td>Data processing as scale</td>
</tr>
</tbody>
</table>


<h4 id="block-2-quantitative-refresher"><strong>Block 2</strong>: Quantitative refresher</h4>






<table class="table table-hover table-success">
<thead>
<tr>
<th>#</th>
<th>What</th>
<th>Why</th>
</tr>
</thead>

<tbody>
<tr>
<td>2.0</td>
<td><a href="/BMMB554/lecture10/"><strong>The data</strong></a></td>
<td>The nature of sequencing data: A quick overview</td>
</tr>

<tr>
<td>2.1</td>
<td><strong>Fundamentals</strong></td>
<td>Probability, Descriptive statistics, Correlation analysis, and Logarithms</td>
</tr>

<tr>
<td>2.2</td>
<td><strong>Statistical analyses</strong></td>
<td>Distributions, Sampling, Significance, Permutation, Bayes Theorem</td>
</tr>

<tr>
<td>2.3</td>
<td><strong>Visualization</strong></td>
<td>Useful versus meaningless</td>
</tr>

<tr>
<td>2.4</td>
<td><strong>Linear algebra</strong></td>
<td>Matrix operations, Eigenvalues and Eigenvectors</td>
</tr>

<tr>
<td>2.5</td>
<td><strong>Discrete data and modeling</strong></td>
<td>Understanding the data and going upward</td>
</tr>

<tr>
<td>2.6</td>
<td><strong>Clustering</strong></td>
<td>Stratifying the data</td>
</tr>

<tr>
<td>2.7</td>
<td><strong>Testing</strong></td>
<td>How to test your hypotheses</td>
</tr>

<tr>
<td>2.8</td>
<td><strong>Multivariate analysis</strong></td>
<td>Finding association among multiple variables</td>
</tr>
</tbody>
</table>


<h4 id="block-3-sources-and-types-of-genomic-data"><strong>Block 3</strong>: Sources and types of genomic data</h4>






<table class="table table-hover table-warning">
<thead>
<tr>
<th>#</th>
<th>What</th>
<th>Why</th>
</tr>
</thead>

<tbody>
<tr>
<td>3.1</td>
<td><strong>DNA (and RNA) sequencing</strong></td>
<td>From Sanger to Nanopores</td>
</tr>

<tr>
<td>3.2</td>
<td><strong>Variation</strong></td>
<td>Finding and interpreting genetic differences</td>
</tr>

<tr>
<td>3.3</td>
<td><strong>Transcriptomics</strong></td>
<td>Measuring gene expression</td>
</tr>

<tr>
<td>3.4</td>
<td><strong>Transcriptomics</strong></td>
<td>Measuring shapes</td>
</tr>

<tr>
<td>3.5</td>
<td><strong>DNA/Protein interactions</strong></td>
<td>Assessing gene regulation and genome architecture</td>
</tr>

<tr>
<td>3.6</td>
<td><strong>Metagenomics</strong></td>
<td>Analysis of complex mixtures</td>
</tr>

<tr>
<td>3.7</td>
<td><strong>About counts</strong></td>
<td>NGS data is count data. There are common themes in read count analysis</td>
</tr>
</tbody>
</table>


<h4 id="block-4-computational-biology-basics"><strong>Block 4</strong>: Computational biology basics</h4>






<table class="table table-hover table-danger">
<thead>
<tr>
<th>#</th>
<th>What</th>
<th>Why</th>
</tr>
</thead>

<tbody>
<tr>
<td>4.1</td>
<td><strong>Alignment</strong></td>
<td>Fundamental concepts, Global and local alignment</td>
</tr>

<tr>
<td>4.2</td>
<td><strong>Aligning many sequences quickly</strong></td>
<td>Mapping in the age of billion-read datasets</td>
</tr>

<tr>
<td>4.3</td>
<td><strong>Bloom filters</strong></td>
<td>Searching in the age of Exobyte databases</td>
</tr>

<tr>
<td>4.4</td>
<td><strong>Assembly</strong></td>
<td>Reconstructing genomes and transcriptomes</td>
</tr>
</tbody>
</table>


<h2 id="final-project">Final Project</h2>

<hr />

<p>We will get into specifics of the final project before spring break.</p>

<h2 id="ecos-teaching-statement">ECoS Teaching Statement</h2>

<hr />

<p>In an examination setting, unless the instructor gives explicit prior instructions to the contrary, violations of academic integrity shall consist of any attempt to receive assistance from written or printed aids, from any person or papers or electronic devices, or of any attempt to give assistance, whether the student doing so has completed his or her own work or not. Other violations include, but are not limited to, any attempt to gain an unfair advantage in regard to an examination, such as tampering with a graded exam or claiming another&rsquo;s work to be one&rsquo;s own. Other assessments (including ANGEL-administered quizzes and assessments as well as homework assignments) are expected to represent your own independent work unless specifically stated otherwise. Failure to comply will lead to sanctions against the student in accordance with the Policy on Academic Integrity in the Eberly College of Science. The Eberly College of Science Code of Mutual Respect and Cooperation (www.science.psu.edu/climate/Code-of-Mutual-Respect-final.pdf) embodies the values that we hope our faculty, staff, and students possess and will endorse to make The Eberly College of Science a place where every individual feels respected and valued, as well as challenged and rewarded.   The Eberly College of Science is committed to the academic success of students enrolled in the College&rsquo;s  courses and undergraduate programs. When in need of help, students can utilize various College and University wide resources for learning assistance (<a href="http://www.science.psu.edu/advising/success">http://www.science.psu.edu/advising/success</a>). Penn State welcomes students with disabilities into the University&rsquo;s educational programs. If you have a disability-related need for reasonable academic adjustments in this course, contact the Office for Disability Services (ODS) at 814-863-1807 (V/TTY). For further information regarding ODS, please visit the Office for Disability Services Web site at <a href="http://equity.psu.edu/ods/.  ">http://equity.psu.edu/ods/.  </a> In order to receive consideration for course accommodations, you must contact ODS and provide documentation (see the <a href="http://equity.psu.edu/student-disability-resources/guidelines">documentation guidelines</a>). If the documentation supports the need for academic adjustments, ODS will provide a letter identifying appropriate academic adjustments. Please share this letter and discuss the adjustments with your instructor as early in the course as possible. You must contact ODS and request academic adjustment letters at the beginning of each semester.</tt></p>



      </main>
    </div>
    <footer id="footer" class="mt-auto text-center text-muted">
  <div class="container">
    Nekrutenko Lab 2019 | CC-BY
  </div>
</footer>
    <script src="https://nekrut.github.io/BMMB554/js/feather.min.js"></script>
<script>
  feather.replace()
</script>


    
  

  </body>
</html>