<!doctype html>
<html lang="en" class="h-100">
  <head>
  <meta name="generator" content="Hugo 0.53" />
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <link rel="stylesheet" href="https://nekrut.github.io/BMMB554/css/bootstrap.min.css">
  
  
  <title>Lecture 4: Python refresher 2 | BMMB554</title>
  <style>
.container {
  max-width: 800px;
}
#nav a {
  font-weight: bold;
  color: inherit;
}
#nav a.nav-link-active {
  background-color: #212529;
  color: #fff;
}
#nav-border {
  border-bottom: 1px solid #212529;
}
#main {
  margin-top: 1em;
  margin-bottom: 4em;
}
#home-jumbotron {
  background-color: inherit;
}
#footer .container {
  padding: 1em 0;
}
#footer a {
  color: inherit;
  text-decoration: underline;
}
.font-125 {
  font-size: 125%;
}
.tag-btn {
  margin-bottom: 0.3em;
}
pre {
  background-color: #f5f5f5;
  border: 1px solid #ccc;
  border-radius: 4px;
  padding: 16px;
}
pre code {
  padding: 0;
  font-size: inherit;
  color: inherit; 
  background-color: transparent;
  border-radius: 0;
}
code {
  padding: 2px 4px;
  font-size: 90%;
  color: #c7254e;
  background-color: #f9f2f4;
  border-radius: 4px;
}
img {
  max-width: 100%;
}
</style>

  <script type="text/javascript" async
  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
  MathJax.Hub.Config({
  tex2jax: {
    inlineMath: [['$','$'], ['\\(','\\)']],
    displayMath: [['$$','$$']],
    processEscapes: true,
    processEnvironments: true,
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
    TeX: { equationNumbers: { autoNumber: "AMS" },
         extensions: ["AMSmath.js", "AMSsymbols.js"] }
  }
  });
  MathJax.Hub.Queue(function() {
    
    
    
    var all = MathJax.Hub.getAllJax(), i;
    for(i = 0; i < all.length; i += 1) {
        all[i].SourceElement().parentNode.className += ' has-jax';
    }
  });

  MathJax.Hub.Config({
  
  TeX: { equationNumbers: { autoNumber: "AMS" } }
  });
</script>
  
</head>
  <body class="d-flex flex-column h-100">
    <div id="nav-border" class="container">
  <nav id="nav" class="nav justify-content-center">
  
  
  
    
    
      
        
      
    
    
    <a class="nav-link " href="/BMMB554/"><i data-feather="home"></i> Home</a>
  
    
    
      
        
      
    
    
    <a class="nav-link " href="/BMMB554/post/"><i data-feather="book-open"></i> Lectures</a>
  
    
    
      
        
      
    
    
    <a class="nav-link " href="/BMMB554/tags/"><i data-feather="tag"></i> Tags</a>
  
    
    
      
        
      
    
    
    <a class="nav-link " href="/BMMB554/about/"><i data-feather="list"></i> Syllabus</a>
  
  </nav>
</div>
    <div class="container">
      <main id="main">
        

<h1>Lecture 4: Python refresher 2</h1>


<i data-feather="calendar"></i> <time datetime="2019-01-16">Jan 16, 2019</time>

  <br>
  <i data-feather="tag"></i>
  
  
  <a class="btn btn-sm btn-outline-dark tag-btn" href="https://nekrut.github.io/BMMB554/tags/python">python</a>
  
  
  <a class="btn btn-sm btn-outline-dark tag-btn" href="https://nekrut.github.io/BMMB554/tags/conditionals">conditionals</a>
  
  
  <a class="btn btn-sm btn-outline-dark tag-btn" href="https://nekrut.github.io/BMMB554/tags/working-with-files">working with files</a>
  
  
  <a class="btn btn-sm btn-outline-dark tag-btn" href="https://nekrut.github.io/BMMB554/tags/tnseq">TnSeq</a>
  

<br><br>


<p><a href="https://xkcd.com/927/"><img src="https://imgs.xkcd.com/comics/standards.png" alt="" /></a></p>

<h1 id="conditionals-loops-and-files">Conditionals, Loops, and Files</h1>

<div class="alert alert-info" role="alert">
  This material is heavily inspired by <a href="http://pycam.github.io/">Materials for the course run by the Graduate School of Life Sciences, University of Cambridge</a> with some modifications.
</div>

<h1 id="let-s-get-started">Let&rsquo;s get started</h1>

<ul>
<li>Point your browser to <a href="http://colab.research.google.com">http://colab.research.google.com</a></li>
<li>Open notebook <code>2019/ipynb/intro2.ipynb</code> from <code>https://github.com/nekrut/BMMB554</code></li>
</ul>

<h1 id="case-study-transposon-insertion-sequencing-tnseq">Case study: Transposon insertion sequencing (TnSeq)</h1>

<p>Transposon insertion allows genome-wide identification of essential genes. The idea is simple = important genes cannot be disrupted and there will not accept any transposon insertions.</p>

<div class="card mb-3">
  <img class="card-img-top" src="/BMMB554/img/nrmicro.2015.7-f1.jpg" alt="Card image cap">
  <div class="card-footer">
        <footer class="blockquote-footer">
            An overview of transposon insertion sequencing from <a href="https://www.nature.com/articles/nrmicro.2015.7">a review</a> by Chao et al. 2016. <hr>
            <b>a</b> | A high-density transposon insertion library containing multiple insertions in every non-essential genomic locus is created and then grown under different conditions (for example, selective (condition B) and non-selective (condition A)), and mutants that are viable under each condition are recovered. The transposon junctions in the selected and non-selected pools are attached to sequencing adaptors and amplified, then subjected to high-throughput sequencing. <b>b</b> | The sequences are mapped to the genome and the read counts at each insertion site are subjected to statistical analysis to define the genomic loci that seem significantly underrepresented under the selective growth condition (conditionally essential loci, blue box). The insertions in the non-selected library (condition A) can also be statistically evaluated to define essential loci (orange box) that are required for growth under optimal conditions. In highly saturated libraries (see high-saturation graph), non-essential and essential genes are easily distinguishable; conversely, many non-essential genes are likely to not have been disrupted by chance in less saturated libraries (see low-saturation graph).
        </footer>
    </div>
</div>

<p><a href="https://bmb.psu.edu/directory/txm50">Tim Meredith</a> from PennState and colleagues developed <a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-015-1361-3">a more advanced version of TnSeq</a> that alters gene expression level.</p>

<div class="card mb-3">
  <img class="card-img-top" src="/BMMB554/img/tnseq_santiago2015.png" alt="Card image cap">
  <div class="card-footer">
        <footer class="blockquote-footer">
            From Santiago et al. <a href="https://doi.org/10.1186/s12864-015-1361-3">2015</a>.<hr>
            Protocol for the preparation of a high quality transposon DNA library for NGS. (A) (1) Genomic DNA is isolated and digested with NotI. High molecular weight DNA is selectively precipitated using an 8%PEG + NaCl solution, and transposon-plasmid junctions (106 bp) are removed in the supernatant. A biotinylated dsDNA adapter with NotI overhang is ligated (2) before digestion with MmeI, which cuts non-specifically 20 bp from its recognition site within ITR2 into the genome to liberate biotinylated-transposon-genome junctions as short DNA fragments (114 bp). (3). Biotinylated fragments are bound to streptavidin beads (4), and an Illumina sequencing primer adapter containing an indexing barcode and MmeI compatible ends is ligated (5). Primers annealing to the P7 site and the Illumina sequencing primer adapter sequence (with a P5 site overhang) are used to PCR amplify the transposon-genome junctions (6), agarose gel purified, and submitted for Illumina sequencing (7). NGS reads capture both the 16-bp of flanking genomic DNA as well as the transposon donor specific barcode located between the P7 and ITR2. (B) Fragments arising from transposon-plasmid junctions are removed by size selective PEG-NaCl precipitation, while the remaining fragments lack both P7 annealing sites and MmeI sites for ligation of the Illumina sequencing primer adapter. These fragments are therefore not amplified in step (6) of 4A. (C) By performing the size-selective precipitation on a 1 kb DNA ladder, we show that small 300 bp fragments of DNA are retained in the solution (SN), while larger DNA is precipitated (P). (D) Six transposon donor constructs were multiplexed and designed to attenuate expression of genes proximal to the insertion site according to the regulatory elements located at the ends of the transposon backbone. Each donor can be identified from NGS reads by the unique 3 bp barcode.
        </footer>
    </div>
</div>

<p>We will review the analysis of TnSeq data in greater detail once we are done with programming and quantitative basics. For now let&rsquo;s assume that the data is already processed and exists in the following format:</p>
<div class="highlight"><pre style=";-moz-tab-size:4;-o-tab-size:4;tab-size:4">NC_007795.1     2500004 2500006 4.0     3.0     1.0     1.0     2.0     10.0
NC_007795.1     2500008 2500010 3.0     11.0    0.0     1.0     2.0     11.0
NC_007795.1     2500020 2500022 4.0     13.0    8.0     2.0     5.0     13.0
NC_007795.1     2500022 2500024 27.0    15.0    8.0     3.0     7.0     18.0
NC_007795.1     2500031 2500033 27.0    13.0    8.0     3.0     6.0     23.0
NC_007795.1     2500033 2500035 26.0    12.0    8.0     3.0     5.0     23.0
NC_007795.1     2500037 2500039 28.0    12.0    6.0     5.0     7.0     16.0
NC_007795.1     2500043 2500045 6.0     11.0    6.0     4.0     5.0     10.0</pre></div>
<p>where columns are as follows:</p>

<ol>
<li>Chromosome ID (in this case simply genome name)</li>
<li>Start cooredinate</li>
<li>End coordinate</li>
<li>Read count for <code>blunt</code></li>
<li>Read count for <code>cap</code></li>
<li>Read count for <code>dual_pen</code></li>
<li>Read count for <code>erm</code></li>
<li>Read count for <code>pen</code></li>
<li>Read count for <code>tuf</code></li>
</ol>

<p>Let&rsquo;s use these data to practice file access operation in python. Files can be found here:</p>

<ul>
<li><a href="/BMMB554/tnseq_data_short.txt">Small file</a></li>
<li><a href="/BMMB554/tnseq_data.txt">Large file</a></li>
</ul>

<p>We will use <code>intro2.ipynb</code> notebook to process these files</p>

<p># Homework</p>

<p><a href="/BMMB554/yeast_genes.txt">Here</a> is a file containing information about yeast genes. It looks like this:</p>
<div class="highlight"><pre style=";-moz-tab-size:4;-o-tab-size:4;tab-size:4">#systematic_name	standard_name	chrom	start	end
21S_rRNA_4	21S_RRNA_4	chrmt	61868	62447
9S_rRNA_1	9S_RRNA_1	chrmt	1	11
9S_rRNA_5	9S_RRNA_5	chrmt	85290	85779
ETS1-1	ETS1-1	chrXII	457733	458432
ETS1-2	ETS1-2	chrXII	466870	467569
ETS2-1	ETS2-1	chrXII	451575	451785
ETS2-2	ETS2-2	chrXII	460712	460922
HRA1	HRA1	chrI	99305	99868
ICR1	ICR1	chrIX	393884	397082
IRT1	IRT1	chrX	605936	607424</pre></div>
<p>For every gene, its location and coordinates are recorded.
You should read through the file and store the data into an appropriate structure.
Then answer these questions:</p>

<iframe src="https://docs.google.com/forms/d/e/1FAIpQLSdoTKg6kj7MHoHlffNBQ2CUXYCzIU1JYJDlse-YFov45_O7AQ/viewform?embedded=true" width="800" height="1630" frameborder="0" marginheight="0" marginwidth="0">Loading...</iframe>



      </main>
    </div>
    <footer id="footer" class="mt-auto text-center text-muted">
  <div class="container">
    Nekrutenko Lab 2019 | CC-BY
  </div>
</footer>
    <script src="https://nekrut.github.io/BMMB554/js/feather.min.js"></script>
<script>
  feather.replace()
</script>


    
  

  </body>
</html>